Libs/DICOM/Widgets/ctkDICOMQueryRetrieveWidget.cpp

```diff
@@ -137,13 +137,27 @@ void ctkDICOMQueryRetrieveWidget::setRetrieveDatabase(QSharedPointer<ctkDICOMDat
   d->RetrieveDatabase = dicomDatabase;
 }
 
+//----------------------------------------------------------------------------
+void ctkDICOMQueryRetrieveWidget::setRetrieveDatabase(ctkDICOMDatabase* dicomDatabase)
+{
+  Q_D(ctkDICOMQueryRetrieveWidget);
+  if (dicomDatabase == nullptr)
+    {
+    logger.error("Failed to set retrieve database - nullptr given");
+    return;
+    }
+
+  d->RetrieveDatabase = QSharedPointer<ctkDICOMDatabase>(dicomDatabase);;
```

Comment:
As I understand, ownership of the database is passed to this widget (the widget will delete the object when it is destroyed). This would be unexpected and have quite high chance of double-delete.

Maybe we could fix this by [setting a custom deleter function](https://doc.qt.io/qt-6/qsharedpointer.html#QSharedPointer-2) that does not actually delete the object (does not do anything).

Reply:
I didn't consider this. What I'm not sure about is the reason to use a QSharedPointer in the first place. I thought about making it a simple pointer but it may break the API and didn't want to do that. Not sure what you think about this.

Reply:
Shared pointer makes sense because the database may be used without a widget or used by multiple widgets; and also because it is very easy to cleanly delete it during application exit. Unfortunately, shared pointer support has not been implemented in PythonQt, so we need workarounds.

Search for `skipDelete` in this repository for how a dual mode use is inplemented in another class.